functions/go/set-namespace/transformer/namespace.go

```diff
@@ -18,206 +18,311 @@ import (
+				continue
 			}
-			o.SetOrDie(&subjects, "subjects")
-		}
-		// "IsNamespaceScoped" detects resource scope by checking whether `metadata.namespace` is set.
-		// This is valid hypothesis is for custom k8s resources. For k8s resources, the function always gives accurate scope information.
-		if o.IsNamespaceScoped() {
-			namespace := o.NestedStringOrDie("metadata", "namespace")
-			visitor(NewNamespace(o, &namespace))
-			o.SetOrDie(&namespace, "metadata", "namespace")
+			if origin.Namespace == fn.UnknownNamespace {
+				// This should rarely happen.
+				return nil, nil, fmt.Errorf("%v is namespace-scoped, but has cluster-scoped or unknown scoepd origin %v",
+					o.ShortString(), origin.String())
+			}
+			originNss.Insert(origin.Namespace)
+		} else {
+			results = append(results, fn.GeneralResult(fmt.Sprintf(
+				"%v does not have upstream origin.", o.ShortString()), fn.Warning))
 		}
 	}
+	return originNss.List(), results, nil
 }
 
-// ReplaceNamespace iterates the `objects` to replace the `OldNs` with `newNs` on namespace field.
-func ReplaceNamespace(objects []*fn.KubeObject, newNs string, count *int, oldNamespace map[string]struct{}) {
-	VisitNamespaces(objects, func(ns *Namespace) {
-		if *ns.Ptr != newNs {
-			oldNamespace[*ns.Ptr] = struct{}{}
-			*ns.Ptr = newNs
-			*count += 1
+// WalkAndReplace iterate each KRM resource and updates the "namespace" fields.
+func WalkAndReplace(objects fn.KubeObjects, newNs string, matchers ...string) (fn.Results, int, []string) {
+	count := 0
+	oldnss := sets.NewString()
+	var results fn.Results
+	VisitAll(objects, func(origin string, currentPtr *string, idStr ...string) {
+		// Skip if the resource is a cluster scoped or unknown scoped resource.
+		if origin == fn.UnknownNamespace {
+			return
+		}
+		if *currentPtr == "" {
+			*currentPtr = fn.DefaultNamespace
+		}
+		if *currentPtr == newNs {
+			return
+		}
+		// matcher not given, update all.
+		change := false
+		if len(matchers) == 0 {
+			change = true
+		} else {
+			for i := range matchers {
+				if matchers[i] == origin {
+					results = append(
+						results, fn.GeneralResult(fmt.Sprintf("%s has matching origin %s", idStr, origin), fn.Info))
+				}
+				if matchers[i] == *currentPtr || matchers[i] == origin {
+					change = true
+				}
+			}
+		}
+		if change {
+			oldnss.Insert(*currentPtr)
+			*currentPtr = newNs
+			count += 1
 		}
 	})
+	return results, count, oldnss.List()
 }
 
-// GetDependsOnMap iterates `objects` to get the annotation which contains namespace value.
-func GetDependsOnMap(objects []*fn.KubeObject) map[string]bool {
-	dependsOnMap := map[string]bool{}
-	VisitNamespaces(objects, func(ns *Namespace) {
-		key := ns.GetDependsOnAnnotation()
-		dependsOnMap[key] = true
-	})
-	return dependsOnMap
+// VisitAll applies "visitor" function to both namespace scoped and cluster scoped resource.
+func VisitAll(objects fn.KubeObjects, visitor func(origin string, currentPtr *string, idStr ...string)) {
```

Comment:
This might be something useful in the SDK after generalization.

Reply:
Totally agree!